### ISSUE TITLE
[Not a Bug] Noob Question - Can anyone explain why I would use this over the standard Microsoft.DependencyInjection Library

### ISSUE THREAD

**Issue by jeffward01** [z4kn4fein/stashbox#109]
Hello all!

Sorry for asking a simple question like this, but I could not find another place to ask this. 

I understand that in itself the Microsoft 'standard' dependency injection library is somewhat of an anti-pattern and also limited in features compared to Autofac for example.

I looked at the code and examples for Stashbox, it looks VERY CLEAN!! I would love to use it, but I would love to know a reason of why I should use it specifically.

For example, if I need event interception of calling methods, or some other fancy things I can use AutoFac.   

What does Stashbox bring to the playing field that makes it stand apart from the rest (aside from a super clean and nice Fluent Interface pattern <3) 

Thanks! 

@z4kn4fein

**Comment by vnvizitiu** [comment]
Hello,

I know this was meant for @z4kn4fein but if I can chime in for my own use cases.

Here are a few reasons that pulled me towards it:
 * the usage of CompositionRoots, as such, each assembly can have its own internal wiring which I can just discover at runtime and use, making swapping in and out assemblies very easy
 * It's integration with NSubstitute for mocking in test assemblies
 * Easy to use in both Console and Asp.Net project (this I think has been addressed by Microsoft already)
 * The "AsImplementedTypes" extension method so that if a class implements more than one interface, I can resgister it under all easily (depending on the scenario)
 * Assembly scanning. As an example, I have a hierarchy of classes and I want everything that implements that base particular base class to be registered so that I only have to wire it up once for that hierarchy and then never have to look at it again, just create a class that inherits that base class and it's already registered at runtime without remembering to add a line of code for it.
 * Best of all for 3rd party libraries, I can convert an IServiceCollection to Stashbox registrations using the "container.RegisterServiceDescriptors(serviceCollection);" method, so for libraries that use the "serviceCollection.Add" approach (like SignalR, Swagger, or even MVC) I can use a normal IServiceCollection and then just register it into Stashbox.
 
All of these features I'm currently using in my projects.

Hope this is helpfull.

Cheers,

**Comment by z4kn4fein** [comment]
Hey,

@vnvizitiu Thank you for sharing your points, it's really good to see how you are using Stashbox, and which features exactly!

@jeffward01 Thank you for reaching out, I'm happy you like Stashbox!
Unfortunately, I can't give you a straight answer for why you should use Stashbox over any other containers, because I think the answer depends on your needs and preferences the most. One of these preferences could be the need for a fluent API that you already mentioned Stashbox has, but there are also more aspects you have to examine first to make the final decision.

It could be a really hard choice because there are a bunch of Dependency Injection container implementations out there that have similar functionalities with different or just slightly different APIs. Also, most of them have a lot of features, that you might even don't know yet whether you want to use ever, or not. This could be overwhelming and also could move the reasons behind the choice to a more emotional level. Like which one feels the most natural to you to use?

I think it's a bit pointless to repeat here the list of features that Stashbox has because you probably already see them in the [docs](https://z4kn4fein.github.io/stashbox/#/). Also, I would like to be completely honest here (because I hate direct marketing 😄), I don't think Stashbox has more features than Autofac, despite I'm not super familiar with it. Stashbox has a different API, a similar feature set, a similar set of extensions, but better performance based on [these](https://danielpalme.github.io/IocPerformance/) metrics. So my key points are that it has a decent amount of features, configuration capabilities, and extensions along with a good performance. Also, I'm trying to do my best in terms of support and feature requests.

Maybe another thing that I can mention here is a story about how Stashbox saved lots of minutes in our ASP.NET Core integration test runs recently. 
In our integ tests at my company, we had to replace some services several times with mocks at runtime, and also we used a mock DB behind our applications. 
We managed to keep our data during the test execution in a single DB instance, but for replacing services multiple times we had to always start a new instance of our ASP.NET Core application to prevent the tests from interfering with each other. Another issue was that after the ServiceProvider is built from the MS DI's ServiceCollection, you are not able to apply changes to the services, however, Stashbox's multitenant functionality is able to do this with child containers. We were able to open a new tenant and override specific services with mocks in every test case that needed it with starting the application itself only once for the whole test assembly. This reduced the test execution time by minutes (I can't recall exactly how many minutes) and gave us a very easy-to-use service separation ability.

I'm not sure this story fits in the context for you or not, it just came into my mind, and thought it might be a good tale for the topic. 😄 
Anyway, in the end, I think the most important thing is knowing your needs and preferences. And of course, the feeling of using the given library.

Please let me know your thoughts, and if you have questions feel free to ask them!

Thanks!

**Comment by z4kn4fein** [comment]
I forgot to mention that I'm playing with a feature that I don't think any other container has (not 100% sure), but I see some sort of need for it. I'm not sure that it's useful in this form, it's kind of an undocumented POC yet.

You can attach asynchronous initializer functions to services like:
```csharp
.Register<Service>(c => c.WithAsyncInitializer((service, resolver, token) => service.InitAsync()));
```
They can be invoked by a scope (only on already instantiated services):
```csharp
await scope.InvokeAsyncInitializers(token);
```
The use case is to support those services that have a dependency on an asynchronous resource state that is ideally enough to read at the initialization phase, but for clear reasons, it can't be done in a synchronous constructor.

The idea is that during an ASP.NET Core controller activation when a controller is already instantiated, but before its action is invoked, we can asynchronously initialize its services through the request's scope. So, we can spare the manual invocation chain of these kinds of calls in the asynchronous action, which is our first asynchronous space to do so. Or we can also spare the `.CreateIfNotExistsAsync()` type of calls at the beginning of every function of a service that manipulates a particular asynchronous resource.
This can be done through an integration that can decorate the controller invocation with this functionality.

This isn't fully forged yet, I'm still just in a playing phase but I thought it worth mentioning here.

**Comment by z4kn4fein** [comment]
One more thing, I'd be grateful if you could let me know which library you ended up with and why. If it's happen to be Stashbox, I'd be also grateful to hear your feedback on it, so I can use your aspects to improve the project. Thanks in advance!
